docs(ref): Expand on MSRV (#14636)
rust-lang/cargo: src/doc/src/reference/rust-version.md

# Rust Version

The `rust-version` field is an optional key that tells cargo what version of the
Rust toolchain you support for your package.

```toml
[package]
# ...
rust-version = "1.56"
```

The Rust version must be a bare version number with at least one component; it
cannot include semver operators or pre-release identifiers. Compiler pre-release
identifiers such as -nightly will be ignored while checking the Rust version.

> **MSRV:** Respected as of 1.56

## Uses

**Diagnostics:**

When your package is compiled on an unsupported toolchain,

> Cargo will provide clearer diagnostics than reporting invalid syntax or missing functionality in `std`.

Reviewer: ```suggestion
Cargo will provide clearer diagnostics about the insufficient toolchain version rather than reporting invalid syntax or missing functionality in the standard library.
```

This affects all build-targets in the package, including binaries, examples, test suites,
benchmarks, etc.

**Development aid:**

> `cargo add` will auto-select the dependency's version requirement to be low enough to be compatible with your `rust-version`.

Reviewer: ```suggestion
`cargo add` will auto-select the dependency's version requirement to be the latest version compatible with your `rust-version`, which may not be the latest version of the dependency. `cargo add` will warn if it selects an older version, and tell you what MSRV would be required to use the latest version.
```

Author: For the second half of that, I feel like that is clear from the first half and would rather not get into the specifics of how we behave.

Reviewer: I think "which may not be the latest version of the dependency" naturally follows but is still helpful reinforcement to help people who aren't familiar with the problem domain wrap their heads around it.

For the latter half, the exact details aren't load-bearing. The main things I'm trying to convey are:
- `cargo add` doesn't do this *silently*
- If you'd rather use the latest version, here's how you can know what MSRV you'd need to use to do that: `cargo add` will tell you

Author: Further tweaked it.

Reviewer: LGTM, thanks!

Each person can come to a reasonably different justification and
working through that discussion can be frustrating for the involved parties.
This will disempower those who would want to avoid that type of conflict,
which is particularly the case for new or casual contributors who either
feel that they are not in a position to raise the question or
that the conflict may hurt the chance of their change being merged.

### Multiple Policies in a Workspace

Cargo allows supporting multiple policies within one workspace.

Verifying specific packages under specific versions can get complicated.
Tools like [`cargo-hack`](https://crates.io/crates/cargo-hack) can help.

For any dependency shared across policies,
the lowest common versions must be used as Cargo
[unifies SemVer-compatible versions](resolver.md#semver-compatibility),
potentially limiting access to features of the shared dependency for the workspace member with the higher `rust-version`.

### One or More Policies

One way to mitigate the downsides of supporting older Rust versions is to apply your policy to older major or minor versions of your package that you continue to support.

> You likely need still need a policy for what Rust versions the development branch support compared to the release branches for those major or minor versions.

Reviewer 2: ```suggestion
You likely still need a policy for what Rust versions the development branch support compared to the release branches for those major or minor versions.
```

Only updating the development branch when "needed"' can help reduce the number of supported release branches.

There is the question of what can be backported into these release branches.
By backporting new functionality between minor versions,
the next available version would be missing it which could be considered a breaking change, violating SemVer.
Backporting changes also comes with the risk of introducing bugs.

Supporting older versions comes at a cost.
This cost is dependent on the risk and impact of bugs within the package and what is acceptable for backporting.
Creating the release branches on-demand and putting the backport burden on the community are ways to balance this cost.

There is not yet a way for dependency management tools to report that the version is still supported,

> shifting the responsibility for users to notice this in documentation.

Reviewer 2: ```suggestion
shifting the responsibility to users to notice this in documentation.
```

# Rust Version

> The `rust-version` field is an optional key that tells cargo what version of the

Reviewer 2: Found this :)

```suggestion
The `rust-version` field is an optional key that tells Cargo what version of the
```

Author: That was mirroed from the manifest page which uses that a lot.  I'm unsure if I want to go down that rabbit hole in this PR.

Example version policies:
- "N-2", meaning "latest version with a 2 release grace window for updating"
- Every even release with a 2 release grace window for updating
- Every version from this calendar year with a one year grace window for updating

> **Note:** To find the minimum `rust-version` compatible with your project as-is, you can use third-party tools like [`cargo-msrv`](https://crates.io/crates/cargo-msrv).

### Update timeline

When your policy specifies you no longer need to support a Rust version, you can update `rust-version` immediately or when needed.

By allowing `rust-version` to drift from your policy,
you offer users more of a grace window for upgrading.
However, this is too unpredictable to be relied on for aligning with the Rust version users track.

The further `rust-version` drifts from your specified policy,
the more likely users are to infer a policy you did not intend,
leading to frustration at the unmet expectations.

When drift is allowed,
there is the question of what is "justifiable enough" to drop supported Versions.

> Each person can come to a reasonably different justification and
> working through that discussion can be frustrating for the involved parties.

Reviewer 2: Better to read.

```suggestion
Each person can come to a reasonably different justification,
and working through that discussion can be frustrating for the involved parties.
```
